src/charm.py

```diff
@@ -2,203 +2,144 @@
-                            "prometheus.io/path": "/metrics",
-                        }
-                    },
+        self.lightkube_client = Client(namespace=self.model.name, field_manager="lightkube")
+
+        self.resource_handler = KRH(
+            template_files=self._template_files,
+            context=self._context,
+            field_manager=self.model.app.name,
+        )
+
+        self._mutating_webhook_name = f"{self.model.app.name}-webhook-config"
+        self._container_name = "spark"
+        self.container = self.unit.get_container(self._container_name)
+
+        self.framework.observe(self.on.spark_pebble_ready, self._on_spark_pebble_ready)
+        self.framework.observe(self.on.config_changed, self._on_config_changed)
+        self.framework.observe(self.on.remove, self._on_remove)
+
+    @property
+    def _template_files(self):
+        src_dir = Path("src")
+        manifests = [file for file in glob.glob(f"{src_dir}/*.yaml")]
+        return manifests
+
+    @property
+    def _context(self):
+        context = {
+            "app_name": self.model.app.name,
+            "model_name": self.model.name,
+        }
+        return context
+
+    @property
+    def _pebble_layer(self):
+        pebble_layer = {
+            "summary": "spark layer",
+            "description": "pebble config layer for spark-k8s",
+            "services": {
+                self._container_name: {
+                    "override": "replace",
+                    "summary": "Spark Operator layer",
+                    "startup": "enabled",
+                    "command": (
+                        f"/usr/bin/tini -s -- /usr/bin/spark-operator -v=2 "
+                        "-logtostderr "
+                        f"-namespace={self.model.name} "
+                        "-enable-ui-service=true "
+                        "-controller-threads=10 "
+                        "-resync-interval=30 "
+                        "-enable-batch-scheduler=false "
+                        "-enable-metrics=true "
+                        "-metrics-labels=app_type "
+                        f"-metrics-port={self.model.config['metrics-port']} "
+                        "-metrics-endpoint=/metrics "
+                        "-enable-resource-quota-enforcement=false "
+                        "-enable-webhook=true "
+                        f"-webhook-svc-namespace={self.model.name} "
+                        f"-webhook-port={self.model.config['webhook-port']} "
+                        f"-webhook-svc-name={self.model.app.name} "
+                        f"-webhook-config-name={self._mutating_webhook_name} "
+                        f"-webhook-namespace-selector=model.juju.is/name={self.model.name} "
+                        "-webhook-fail-on-error=true"
+                    ),
                 }
             },
+        }
+        return pebble_layer
+
+    def _on_spark_pebble_ready(self, _):
```

Comment:
We might want to follow the pattern of:

```python
if not container.can_connect():
    defer()
new_layer = _pebble_layer
current_layer = container.get_plan
if new_layer != current_layer:
    container.add_layer()
    container.replan()
```

Reply:
What's the benefit of comparing the current and new layer? Does pebble not check before replanning?

Reply:
Replan will start changed services and startup those who are startup-enbaled ones. We run this only after adding a new layer, and when we add a new layer we want to make sure we are adding one that actually changes the config, hence the comparison between current and new layer.